### PR TITLE
edf_schedule: improve algorithm to use dynamic deadlines

### PR DESCRIPTION
**Pull request by tlauda**
Improves EDF scheduling algorithm to use dynamically calculated
deadlines instead of static ones. It also simplifies the way
it chooses the next task to be executed and is first step for
further simplifications.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

File: src/drivers/imx/ipc.c [hunk 1]
```diff
@@ -171,6 +171,10 @@ struct ipc_data_host_buffer *ipc_platform_get_host_buffer(struct ipc *ipc)
 
 int platform_ipc_init(struct ipc *ipc)
 {
+	struct task_ops ops = {
+		.run = ipc_platform_do_cmd,
+		.complete = ipc_platform_complete_cmd
+	};
```
Comment (paulstelian97) [inline src/drivers/imx/ipc.c]: I don't like seeing pointers to on-stack structures being passed to functions, if the called function forgets to make a copy of the structure (this is supposedly something persistent) disaster might ensue. I may be tired but I'm not fully certain I can see whether the contents (actual function pointers) are copied or not (they probably are but I just want to make sure).

File: src/ipc/ipc.c [hunk 2]
```diff
@@ -499,3 +499,10 @@ int ipc_init(struct sof *sof)
 
 	return platform_ipc_init(sof->ipc);
 }
+
+struct task_ops ipc_task_ops = {
+	.run		= ipc_platform_do_cmd,
+	.complete	= ipc_platform_complete_cmd,
+	.get_deadline	= ipc_task_deadline,
+	.is_ready	= NULL
+};
```
Comment (paulstelian97) [inline src/ipc/ipc.c]: This commit feels tangential to me and not directly related to dynamic deadlines but rather just a simple refactor. I can accept this but it's something to keep in mind.

File: src/drivers/imx/ipc.c (hunk 1 above)
Comment (tlauda) [reply src/drivers/imx/ipc.c]: They are copied. It's "public" API, so I don't see any problem in simple trusting the function.

Comment (paulstelian97) [reply src/drivers/imx/ipc.c]: I suppose, I just considered this may be potential for future bugs (if somehow a new scheduler will be added and it doesn't account for this in its own init function).

File: src/include/sof/schedule/task.h [hunk 3]
```diff
@@ -43,6 +43,7 @@ enum task_state {
 
 struct task {
 	uint64_t start;
+	uint64_t period;
```
Comment (lgirdwood) [inline src/include/sof/schedule/task.h]: Complexity is starting to grow here, can you comment all these fields now to help other understand the logic.

File: src/schedule/edf_schedule.c [hunk 4]
```diff
@@ -74,30 +70,18 @@ static void edf_scheduler_run(void *data)
...
 			continue;
 
-		edf_pdata = edf_sch_get_pdata(task);
+		deadline = task_get_deadline(task);
 
-		deadline = edf_pdata->deadline;
-
-		if (current >= deadline &&
-		    !(task->flags & SOF_SCHEDULE_FLAG_IDLE)) {
+		if (deadline == SOF_TASK_DEADLINE_NOW) {
```
Comment (lyakh) [inline src/schedule/edf_schedule.c]: this change also makes .flags unused, maybe it's worth at least mentioning this in the commit message. Not hugely important though, just for easier understanding

File: src/schedule/task.c [hunk 5]
```diff
@@ -69,7 +69,8 @@ void task_main_init(void)
 	struct task_ops ops = {
 		.run = main_main,
 		.get_deadline = task_main_deadline,
-		.complete = NULL
+		.complete = NULL,
+		.is_ready = NULL
```
Comment (lyakh) [inline src/schedule/task.c]: again - using a comma also in the last line helps. Also, there's no need to assign NULL to struct members during initialisation - that's done automatically. Can be removed throughout the patch set

File: src/audio/kpb.c [hunk 6]
```diff
@@ -165,7 +165,7 @@ static struct comp_dev *kpb_new(struct sof_ipc_comp *comp)
 			       &ops, /* task ops */
 			       &kpb->draining_task_data, /* task private data */
 			       0, /* core on which we should run */
-			       SOF_SCHEDULE_FLAG_IDLE);
+			       0); /* no flags */
```
Comment (lyakh) [inline src/audio/kpb.c]: Personally I'd rather not keep unused function parameters and struct fields unless there are *certain* plans to begin using them in the near future. Do you have anything in the pipeline to start using .flags? If not, maybe better remove them.

File: src/schedule/edf_schedule.c [hunk 7]
```diff
@@ -71,8 +71,7 @@ static void edf_scheduler_run(void *data)
 		task = container_of(tlist, struct task, list);
 
 		if (task->state != SOF_TASK_STATE_QUEUED &&
-		    task->state != SOF_TASK_STATE_RUNNING &&
-		    task->state != SOF_TASK_STATE_RESCHEDULE)
+		    task->state != SOF_TASK_STATE_RUNNING)
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/schedule/edf_schedule.c]: Yes, exactly.

File: src/audio/kpb.c [hunk 8]
```diff
@@ -150,14 +151,14 @@ static struct comp_dev *kpb_new(struct sof_ipc_comp *comp)
...
-			   0, /* core on which we should run */
-			   SOF_SCHEDULE_FLAG_IDLE);
+	schedule_edf_task_init(&kpb->draining_task, /* task structure */
+			       /* almost idle priority */
+			       SOF_TASK_PRI_ALMOST_IDLE,
+			       kpb_draining_task, /* task function */
+			       NULL, /* no complete function */
+			       &kpb->draining_task_data, /* task private data */
+			       0, /* core on which we should run */
+			       SOF_SCHEDULE_FLAG_IDLE);
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/kpb.c]: Almost all of the rest of the generic scheduling operations are still valid. I would say initializing specific scheduling tasks with never used parameters are pointless and leads only to confusion. You need to specify task type anyway, so using dedicated task init function is a small price for improved clarity.

File: src/include/sof/schedule/task.h [hunk 9]
```diff
@@ -83,6 +84,13 @@ static inline void task_complete(struct task *task)
 		task->ops.complete(task->data);
 }
 
+static inline uint64_t task_get_deadline(struct task *task)
+{
+	assert(task->ops.get_deadline);
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/include/sof/schedule/task.h]: This function is used in separate commit along with the addition of implementations.

File: src/schedule/edf_schedule.c (hunk 4 above)
Comment (tlauda) [reply src/schedule/edf_schedule.c]: I've already noted this in the commit, which removes flags.

File: src/audio/kpb.c (hunk 6 above)
Comment (tlauda) [reply src/audio/kpb.c]: There are specific plans to use flags e.g. to differentiate sporadic tasks.
Better to keep them here.

File: src/schedule/task.c (hunk 5 above)
Comment (tlauda) [reply src/schedule/task.c]: Done.

File: src/drivers/intel/cavs/ipc.c [hunk 10]
```diff
@@ -282,7 +282,8 @@ int platform_ipc_init(struct ipc *ipc)
 	int irq;
 	struct task_ops ops = {
 		.run = ipc_platform_do_cmd,
-		.complete = ipc_platform_complete_cmd
+		.complete = ipc_platform_complete_cmd,
+		.get_deadline = ipc_task_deadline
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/drivers/intel/cavs/ipc.c]: Done.

File: src/include/sof/schedule/edf_schedule.h [hunk 11]
```diff
@@ -36,8 +36,7 @@ struct edf_task_pdata {
 int scheduler_init_edf(void);
 
 int schedule_edf_task_init(struct task *task, uint16_t priority,
-			   enum task_state (*run)(void *data),
-			   void (*complete)(void *data), void *data,
-			   uint16_t core, uint32_t flags);
+			   struct task_ops *ops, void *data, uint16_t core,
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/include/sof/schedule/edf_schedule.h]: Done.

File: src/include/sof/schedule/task.h (hunk 3 above)
Comment (tlauda) [reply src/include/sof/schedule/task.h]: Done.

File: src/include/sof/schedule/task.h [hunk 12]
```diff
@@ -41,18 +42,26 @@ enum task_state {
 	SOF_TASK_STATE_RESCHEDULE,
```
Comment (jajanusz) [inline src/include/sof/schedule/task.h]: Do we still need `SOF_TASK_STATE_RESCHEDULE`, I see you removed it from EDF. Some functions still return that value, but do LL do anything with that?

Comment (tlauda) [reply src/include/sof/schedule/task.h]: Yes, it's used by LL scheduler and also will be used by EDF in the near future.

File: src/schedule/ll_schedule.c [hunk 13]
```diff
@@ -288,46 +280,16 @@ int schedule_ll_task_init(struct task *task, uint16_t type, uint16_t priority,
...
-
-	/* flush for slave core */
-	if (cpu_is_slave(task->core))
-		dcache_writeback_invalidate_region(ll_pdata,
-						   sizeof(*ll_pdata));
-
-	ll_sch_set_pdata(task, ll_pdata);
-
-	return 0;
+	return schedule_task_init(task, type, priority, run, data, core, flags);
```
Comment (ranj063) [inline src/schedule/ll_schedule.c]: @tlauda its not clear the above change is directly because of adding period to struct task. If not, it should be a separate patch?

File: src/include/sof/schedule/schedule.h [hunk 14]
```diff
@@ -50,7 +51,6 @@ enum {
 struct scheduler_ops {
 	void (*schedule_task)(void *data, struct task *task, uint64_t start,
 			      uint64_t period);
-	int (*schedule_task_init)(void *data, struct task *task);
```
Comment (ranj063) [inline src/include/sof/schedule/schedule.h]: @tlauda this change doesnt look complete. If you remove the init from the ops, the free needs to go too right?

File: src/drivers/intel/baytrail/ipc.c [hunk 15]
```diff
@@ -172,6 +172,10 @@ int platform_ipc_init(struct ipc *ipc)
 {
 	struct ipc_data *iipc;
 	uint32_t imrd, dir, caps, dev;
+	struct task_ops ops = {
+		.run = ipc_platform_do_cmd,
+		.complete = ipc_platform_complete_cmd,
+	};
```
Comment (ranj063) [inline src/drivers/intel/baytrail/ipc.c]: @tlauda ipc_platform_do_cmd and ipc_platform_complete_cmd are already abstracted for each platform. Does it make sense to move this struct definition to the ipc header file?

File: src/include/sof/drivers/ipc.h [hunk 16]
```diff
@@ -124,6 +124,14 @@ struct ipc {
 #define ipc_get_drvdata(ipc) \
 	((ipc)->private)
 
+static inline uint64_t ipc_task_deadline(void *data)
+{
+	/* TODO: Currently it's a workaround to execute IPC tasks ASAP.
+	 * In the future IPCs should be considered a budget tasks and
+	 * deadline should be calculated based on that assumption.
```
Comment (kv2019i) [inline src/include/sof/drivers/ipc.h]: @tlauda "should be considered a budget tasks" does not make sense, can you elaborate a bit more?

File: src/include/sof/schedule/task.h [hunk 17]
```diff
@@ -26,6 +26,11 @@ struct sof;
 #define SOF_TASK_PRI_IPC	SOF_TASK_PRI_HIGH
 #define SOF_TASK_PRI_IDC	SOF_TASK_PRI_HIGH
 
+/* predefined EDF task deadlines */
+#define SOF_TASK_DEADLINE_IDLE		UINT64_MAX
+#define SOF_TASK_DEADLINE_ALMOST_IDLE	(SOF_TASK_DEADLINE_IDLE - 1)
```
Comment (kv2019i) [inline src/include/sof/schedule/task.h]: Hmm, this pops out a bit. "almost-idle" smells of implementing a priority scheduler on top of deadline scheduler. Do we have real need for this?

File: src/schedule/edf_schedule.c [hunk 18]
```diff
@@ -74,30 +70,18 @@ static void edf_scheduler_run(void *data)
...
 		}
 
-		delta = deadline - current;
-
-		/* get highest priority */
-		if (task->priority < priority_next) {
-			priority_next = task->priority;
-			delta_next = delta;
+		/* get earliest deadline */
+		if (deadline < deadline_next || !task_next) {
```
Comment (ranj063) [inline src/schedule/edf_schedule.c]: @tlauda is the check for !task_next really needed? If it were ever not NULL, we'd have quit the loop no?

File: src/drivers/intel/cavs/ipc.c [hunk 19]
```diff
@@ -284,6 +284,7 @@ int platform_ipc_init(struct ipc *ipc)
 		.run = ipc_platform_do_cmd,
 		.complete = ipc_platform_complete_cmd,
 		.get_deadline = ipc_task_deadline,
+		.is_ready = NULL,
```
Comment (ranj063) [inline src/drivers/intel/cavs/ipc.c]: @tlauda is it absolutely necessary to set this to NULL? Isnt it a given if the op is not set?

File: src/drivers/imx/ipc.c [hunk 20]
```diff
@@ -171,12 +171,6 @@ struct ipc_data_host_buffer *ipc_platform_get_host_buffer(struct ipc *ipc)
 
 int platform_ipc_init(struct ipc *ipc)
 {
-	struct task_ops ops = {
-		.run = ipc_platform_do_cmd,
-		.complete = ipc_platform_complete_cmd,
-		.get_deadline = ipc_task_deadline,
-		.is_ready = NULL,
-	};
```
Comment (ranj063) [inline src/drivers/imx/ipc.c]: @tlauda this is getting too confusing. You add this struct in one of the patches in the series only to be removed later on?
I think a better way of doing this is to use the common ipc task ops struct from the beginning

File: src/include/sof/schedule/task.h [hunk 21]
```diff
@@ -49,18 +49,19 @@ struct task_ops {
...
-	void *private;
-	struct task_ops ops;
+	uint64_t start;		/**< start time */
+	uint64_t period;	/**< task period */
+	uint16_t type;		/**< type of the task (LL or EDF) */
+	uint16_t priority;	/**< priority of the task (used by LL) */
+	uint16_t core;		/**< execution core */
+	uint16_t flags;		/**< custom flags */
+	enum task_state state;	/**< current state */
+	void *data;		/**< custom data passed to ops */
```
Comment (ranj063) [inline src/include/sof/schedule/task.h]: @tlauda which ops?

File: src/include/sof/schedule/task.h [hunk 22]
```diff
@@ -91,6 +92,14 @@ static inline uint64_t task_get_deadline(struct task *task)
...
 }
 
+static inline bool task_is_ready(struct task *task)
+{
+	if (task->ops.is_ready)
+		return task->ops.is_ready(task->data);
+
+	return true;
+}
+
```
Comment (ranj063) [inline src/include/sof/schedule/task.h]: @tlauda I do not see any users of is_ready in this series. So maybe we can drop this for now and add when needed?

File: src/include/sof/schedule/schedule.h (hunk 14 above)
Comment (tlauda) [reply src/include/sof/schedule/schedule.h]: No, free can still be abstracted.

File: src/drivers/intel/baytrail/ipc.c (hunk 15 above)
Comment (tlauda) [reply src/drivers/intel/baytrail/ipc.c]: Done.

File: src/schedule/edf_schedule.c (hunk 18 above)
Comment (tlauda) [reply src/schedule/edf_schedule.c]: Done.

File: src/include/sof/drivers/ipc.h (hunk 16 above)
Comment (tlauda) [reply src/include/sof/drivers/ipc.h]: Explained it in more details.

File: src/include/sof/schedule/task.h (hunk 17 above)
Comment (tlauda) [reply src/include/sof/schedule/task.h]: Idle task is the main FW loop, so it should be the last task to be executed. Almost idle is deadline dedicated for tasks, which do not have specific requirements to be executed by the end of deadline, but still should be executed before firmware's idle task.

File: src/drivers/intel/cavs/ipc.c (hunk 19 above)
Comment (tlauda) [reply src/drivers/intel/cavs/ipc.c]: Postponed.

File: src/drivers/imx/ipc.c (hunk 20 above)
Comment (tlauda) [reply src/drivers/imx/ipc.c]: Done.

File: src/include/sof/schedule/task.h (hunk 21 above)
Comment (tlauda) [reply src/include/sof/schedule/task.h]: Updated.

File: src/include/sof/drivers/ipc.h (hunk 16 above)
Comment (kv2019i) [reply src/include/sof/drivers/ipc.h]: @tlauda The "considered a budget tasks" still bugs a bit. Are IPCs budget tasks, or should IPCs be allocated a cycle budget? I'm guessing you mean the latter, but the text could be clearer.

File: src/include/sof/drivers/ipc.h [hunk 23]
```diff
@@ -126,6 +126,17 @@ struct ipc {
...
 extern struct task_ops ipc_task_ops;
 
+static inline uint64_t ipc_task_deadline(void *data)
+{
+	/* TODO: Currently it's a workaround to execute IPC tasks ASAP.
+	 * In the future IPCs should be considered a budget tasks and
+	 * deadline should be calculated based on that assumption.
+	 * This means every IPC should have its own maximum number of
+	 * cycles that is required to finish processing. Based on that
+	 * value we will be able to calculate max deadline allowed.
```
Comment (kv2019i) [inline src/include/sof/drivers/ipc.h]: This is good thanks. A minor nit, "s/to calculate max deadline allowed/to calculate task deadline/".

File: src/include/sof/schedule/task.h (hunk 17 above)
Comment (kv2019i) [reply src/include/sof/schedule/task.h]: @tlauda Ack. It seems this is used quite widely already before the change, so ok.

File: src/audio/kpb.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/kpb.c]: @tlauda sorry, I still disagree. I think this is rather common, that some implementations of a generic function don't use all its parameters, nothing terribly confusing there. Moreover, in such implementations you can mark those unused arguments as __attribute__((unused)) to avoid compiler warnings and simultaneously to document, that that is on purpose.

[the remark this answers is not in the document]
Comment (tlauda) [reply src/audio/kpb.c]: It's just not practical in this case and doesn't benefit us at all. You still need to decide what type of task you want to create. It's the same mechanism as in the object oriented programming languages, where you create specific object and use it based on its generic type. Giving that EDF scheduler is still growing and will require e.g. more operations then simple run method, it is highly unpractical to enforce other schedulers to take task_ops struct.

File: src/include/sof/drivers/ipc.h (hunk 23 above)
Comment (tlauda) [reply src/include/sof/drivers/ipc.h]: Done.

File: src/audio/kpb.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lyakh) [reply src/audio/kpb.c]: @tlauda I think this is similar to scheduling policies in Linux / POSIX. You have a generic scheduling API and you select which scheduling policy you want to govern your task, which will select the respective algorithm. I understand that in some cases generic APIs are a hindrance, they make you drag all the numerous users of that API with you for each change. E.g. that was the case when I was reworking the interrupt handling in SOF. So, this my comment obviously isn't a blocker, if everyone else is fine with this. Particularly this might make sense if you have plans to further work on the EDF scheduler and on the API. Maybe once it's stabilised we can re-unify them.

[the remark this answers is not in the document]
Comment (kv2019i) [reply src/audio/kpb.c]: @tlauda @lyakh I'd lean towards having the separate create functions. As long as the rest of the API is aligned (across schedulers), having a different create function seems ok. Based on @tlauda your earlier comment, this is the intent.
 
OTOH, in Linux kernel, the EDF scheduler is exposed via the standard API (sched_setscheduler takes a sched_param struct that has additional fields for EDF) so it is posssible to use the same interface as well.

**Comment by tlauda** [comment]
@lyakh  @lgirdwood PR ready to be merged. Please re-review.

File: src/audio/kpb.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (lgirdwood) [reply src/audio/kpb.c]: Overall I would like to see consistent APIs for scheduling and all core RTOS services across SOF. Yes, I do see the point to reduce the number of arguments (that are optional for some users while mandatory for others), but keeping APIs consistent guarantees simpler service/code reuse for future use case support (that have some differentiation to current services).
I have no problem here if schedule_task_init_edf() and schedule_task_init_ll() (note the naming change) are helper APIs that simply populate struct task before calling generic schedule_task_init().

File: src/include/sof/schedule/edf_schedule.h [hunk 24]
```diff
@@ -35,4 +35,9 @@ struct edf_task_pdata {
 
 int scheduler_init_edf(void);
 
+int schedule_edf_task_init(struct task *task, uint16_t priority,
```
Comment (lgirdwood) [inline src/include/sof/schedule/edf_schedule.h]: Better to make these helpers (as discussed above) and use "edf" as the suffix (to be consistent with the other APIs).

File: src/include/sof/schedule/task.h [hunk 25]
```diff
@@ -41,6 +42,11 @@ enum task_state {
 	SOF_TASK_STATE_RESCHEDULE,
 };
 
+struct task_ops {
+	enum task_state (*run)(void *data);
+	void (*complete)(void *data);
+};
```
Comment (lgirdwood) [inline src/include/sof/schedule/task.h]: Can we doxygen all these members and structs.

File: src/include/sof/schedule/edf_schedule.h (hunk 24 above)
Comment (tlauda) [reply src/include/sof/schedule/edf_schedule.h]: @lgirdwood I would like to clarify this helper usage. Do you mean we should first call schedule_task_init_edf/ll and after that generic schedule_task_init?

Comment (lgirdwood) [reply src/include/sof/schedule/edf_schedule.h]: yes, this keeps the core APIs consistent whilst allowing some scheduler specific helper APIs.

Comment (tlauda) [reply src/include/sof/schedule/edf_schedule.h]: But schedule_task_init_edf/ll call generic schedule_task_init inside them. I don't think calling it outside of the specific init functions is good.

File: src/include/sof/schedule/task.h (hunk 25 above)
Comment (tlauda) [reply src/include/sof/schedule/task.h]: Done.

**Comment by tlauda** [comment]
@lgirdwood Changed names and added more doxygen comments. Still waiting for clarification on task inits.

File: src/include/sof/schedule/edf_schedule.h (hunk 24 above)
Comment (lgirdwood) [reply src/include/sof/schedule/edf_schedule.h]: Sorry, I meant call inside which is what you are doing.

**Comment by lgirdwood** [comment]
CI usual suspects
